### PR TITLE
ctc loss using dropout layer since .eval() will not work for F.dropout

### PR DESCRIPTION
**Pull request by zh794390558**
since .eval() will not work for F.dropout

### CODE REVIEW

**Comment by sw005320** [comment]
I see.
Do you mean that our (CTC) method would be random due to this?

**Comment by pengchengguo** [comment]
Sorry, we didn't notice this problem before. Thanks, @zh794390558!
We should use `nn.Dropout(p=0.5, inplace=False)` instead of `nn.functional.dropout(input, p=0.5, training=True, inplace=False)`. Could you please update other codes as @Emrys365 mentioned.

I also pasted a related discussion of PyTorch: https://github.com/pytorch/pytorch/issues/26338.

**Comment by sw005320** [comment]
Thanks!
I actually had several reports about the randomness...
I think this way solves the issue and also improves the performance.

**Comment by sw005320** [comment]
@pengchengguo, could you take over this PR?
I think it should improve the performance in some cases I believe.

**Comment by pengchengguo** [comment]
> @pengchengguo, could you take over this PR?
> I think it should improve the performance in some cases I believe.

Sure. I will take over it. Besides, I am also thinking about the compatibility problems since our pretrained models (use `F.dropout()`) don't have a module named `self.dropout`.

**Comment by sw005320** [comment]
Thanks.
And year this does not sound easy...
Can you try to identify where we are using `F.dropout()`?
I just did a quick search (https://github.com/espnet/espnet/search?q=%3Dself.dropout_rate), and we might have some issues in espnet2 ctc, lightconv/dynamic conv (@yuyfujit), GAN TTS wavenet, and tacotron2 decoder (@kan-bayashi)
I think they can be changed since these are new or not default setups.

**Comment by kan-bayashi** [comment]
Tacotron2's dropout is intentional. Not need to change.
And the others seems conditioned on `self.training` so it might not be changed.

**Comment by kamo-naoyuki** [comment]
```
% git grep -n F.dropout
espnet/lm/chainer_backend/lm.py:215:            c[0], h[0] = self.rnn[0](state["c"][0], state["h"][0], F.dropout(emb))
espnet/lm/chainer_backend/lm.py:218:                    state["c"][n], state["h"][n], F.dropout(h[n - 1])
espnet/lm/chainer_backend/lm.py:228:            h[0] = self.rnn[0](state["h"][0], F.dropout(emb))
espnet/lm/chainer_backend/lm.py:238:                h[n] = self.rnn[n](state["h"][n], F.dropout(h[n - 1]))
espnet/lm/chainer_backend/lm.py:240:        y = self.lo(F.dropout(h[-1]))
espnet/nets/chainer_backend/ctc.py:47:            F.dropout(F.pad_sequence(hs), ratio=self.dropout_rate), n_batch_axes=2
espnet/nets/chainer_backend/ctc.py:124:            F.dropout(F.pad_sequence(hs), ratio=self.dropout_rate), n_batch_axes=2
espnet/nets/chainer_backend/transformer/attention.py:95:        p_attn = F.dropout(self.attn, self.dropout)
espnet/nets/chainer_backend/transformer/ctc.py:50:            F.dropout(F.pad_sequence(hs), ratio=self.dropout_rate), n_batch_axes=2
espnet/nets/chainer_backend/transformer/ctc.py:136:            F.dropout(hs, ratio=self.dropout_rate), n_batch_axes=2
espnet/nets/chainer_backend/transformer/decoder_layer.py:71:        e = e + F.dropout(n_e, self.dropout)
espnet/nets/chainer_backend/transformer/decoder_layer.py:75:        e = e + F.dropout(n_e, self.dropout)
espnet/nets/chainer_backend/transformer/decoder_layer.py:79:        e = e + F.dropout(n_e, self.dropout)
espnet/nets/chainer_backend/transformer/embedding.py:37:        return F.dropout(e, self.dropout)
espnet/nets/chainer_backend/transformer/encoder_layer.py:55:        e = e + F.dropout(n_e, self.dropout)
espnet/nets/chainer_backend/transformer/encoder_layer.py:59:        e = e + F.dropout(n_e, self.dropout)
espnet/nets/chainer_backend/transformer/positionwise_feed_forward.py:65:        e = F.dropout(self.act(self.w_1(e)), self.dropout)
espnet/nets/pytorch_backend/rnn/encoders.py:90:                xs_pad = torch.tanh(F.dropout(xs_pad, p=self.dropout))
espnet/nets/pytorch_backend/tacotron2/decoder.py:146:            x = F.dropout(self.prenet[i](x), self.dropout_rate)
espnet/nets/pytorch_backend/transformer/dynamic_conv.py:93:        weight = F.dropout(weight, self.dropout_rate, training=self.training)
espnet/nets/pytorch_backend/transformer/dynamic_conv2d.py:105:        weight = F.dropout(weight, self.dropout_rate, training=self.training)
espnet/nets/pytorch_backend/transformer/lightconv.py:94:        weight = F.dropout(self.weight, self.dropout_rate, training=self.training)
espnet/nets/pytorch_backend/transformer/lightconv2d.py:95:        weight_f = F.dropout(weight_f, self.dropout_rate, training=self.training)
espnet/nets/pytorch_backend/transformer/lightconv2d.py:105:        weight = F.dropout(self.weight, self.dropout_rate, training=self.training)
espnet2/asr/ctc.py:132:        ys_hat = self.ctc_lo(F.dropout(hs_pad, p=self.dropout_rate))
espnet2/gan_tts/wavenet/residual_block.py:135:        x = F.dropout(x, p=self.dropout_rate, training=self.training)
```

```
% git grep -n functional.dropout
espnet/nets/pytorch_backend/transducer/transducer_tasks.py:187:            torch.nn.functional.dropout(
```

Excluding tacotron, GAN TTS, and chainer, we just need to fix the following

```
espnet/nets/pytorch_backend/rnn/encoders.py:90:                xs_pad = torch.tanh(F.dropout(xs_pad, p=self.dropout))
espnet2/asr/ctc.py:132:        ys_hat = self.ctc_lo(F.dropout(hs_pad, p=self.dropout_rate))
espnet/nets/pytorch_backend/transducer/transducer_tasks.py:187:            torch.nn.functional.dropout(
```
